### PR TITLE
timeouts: Redesign and unify timeouts

### PR DESCRIPTION
**Pull request by kosstennbl**
## Description
Add common function to run a code block with a timeout. Add several global timeouts for different tasks, which can be configured via environment variables.
Several small fixes and refactoring.

## Issues:
Refs: #335

## How has this been tested:
build, test functionality on increase_decrease_capacity test.

## Types of changes:
Bug-fix, redesign

## Checklist:
**Documentation**
No updates required, ENV variables names could be added somewhere, but the documentation doesn't keep these.

**Issue**
- [ ] Tasks in issue are checked off

### CODE REVIEW

File: src/tasks/kind_setup.cr [hunk 1]
```diff
@@ -116,7 +116,7 @@ class KindManager
     manager = KindManager.new
     cluster = manager.create_cluster(name, kind_config, offline)
     Helm.install("#{name}-plugin #{chart_opts} --namespace kube-system --kubeconfig #{cluster.kubeconfig}")
-    cluster.wait_until_pods_ready()
+    cluster.wait_until_pods_ready
```
Comment (martin-mat) [inline src/tasks/kind_setup.cr:119]: I'd suggest to keep
cluster.wait_until_pods_ready()
to be clear that it is a method call

File: src/tasks/platform/resilience.cr [hunk 2]
```diff
@@ -34,59 +34,51 @@ namespace "platform" do
       KubectlClient::Apply.file("reboot_daemon_pod.yml")
       KubectlClient::Get.wait_for_install("node-failure-coredns")
 
-      pod_ready = ""
-      pod_ready_timeout = 45
       begin
-        until (pod_ready == "true" || pod_ready_timeout == 0)
+
+        execution_complete = repeat_with_timeout(timeout: POD_READINESS_TIMEOUT, errormsg: "Pod daemon installation has timed-out") do
           pod_ready = KubectlClient::Get.pod_status("reboot", "--field-selector spec.nodeName=#{worker_node}").split(",")[2]
```
Comment (martin-mat) [inline src/tasks/platform/resilience.cr:40]: dtto as above, string vs boolean

File: src/tasks/platform/resilience.cr [hunk 3]
```diff
@@ -34,59 +34,51 @@ namespace "platform" do
...
         reboot_daemon_pod = KubectlClient::Get.pod_status("reboot", "--field-selector spec.nodeName=#{worker_node}").split(",")[0]
         start_reboot = KubectlClient.exec("#{reboot_daemon_pod} touch /tmp/reboot")
 
         #Watch for Node Failure.
-        pod_ready = ""
-        node_ready = ""
-        node_failure_timeout = 30
-        until (pod_ready == "false" || node_ready == "False" || node_ready == "Unknown" || node_failure_timeout == 0)
+        execution_complete = repeat_with_timeout(timeout: GENERIC_OPERATION_TIMEOUT, errormsg: "Node shut-off has timed-out") do
           pod_ready = KubectlClient::Get.pod_status("node-failure").split(",")[2]
```
Comment (martin-mat) [inline src/tasks/platform/resilience.cr:55]: string/boolean

File: src/tasks/platform/resilience.cr [hunk 4]
```diff
@@ -34,59 +34,51 @@ namespace "platform" do
...
+          next CNFManager::TestcaseResult.new(CNFManager::ResultStatus::Failed, "Node failed to go offline")
         end
 
         #Watch for Node to come back online
-        pod_ready = ""
-        node_ready = ""
-        node_online_timeout = 300
-        until (pod_ready == "true" && node_ready == "True" || node_online_timeout == 0)
+        execution_complete = repeat_with_timeout(timeout: NODE_READINESS_TIMEOUT, errormsg: "Node startup has timed-out") do
           pod_ready = KubectlClient::Get.pod_status("node-failure", "").split(",")[2]
```
Comment (martin-mat) [inline src/tasks/platform/resilience.cr:69]: string/boolean

File: src/tasks/workload/compatibility.cr [hunk 5]
```diff
@@ -601,7 +582,7 @@ def setup_cilium_cluster(cluster_name : String, offline : Bool) : KindManager::C
     Helm.install("#{cluster_name}-plugin #{chart} #{chart_opts.join(" ")} --namespace kube-system --kubeconfig #{cluster.kubeconfig}")
   end
 
-  cluster.wait_until_pods_ready()
+  cluster.wait_until_pods_ready
```
Comment (martin-mat) [inline src/tasks/workload/compatibility.cr:585]: keep ()

**Review by martin-mat — CHANGES_REQUESTED** [review]

**Comment by martin-mat** [comment]
Additionally, it would be good to document new environment vars to USAGE.md

**Comment by daniel-wilmes** [comment]
@martin-mat I have verified that the fix for node drain works as a single test and in the cert command. However, I will say the cert command does not ever finish, which appears to be a seperate issue from node drain. It may pertain to either sig_term_handled, zombie_handled, or specialized_init_system. The logging doesn't appear to indicate where we are stuck. But for this ticket I think the fix for node_drain should go in.

`---
name: cnf testsuite
testsuite_version: node-drain-fix-2024-05-15-142132-3258c691
status: 
command: ~/.mtx/konstruxx/working/tests/testCHF/cnf-testsuite cert essential
points: 100
exit_code: 0
items:
- name: increase_decrease_capacity
  status: passed
  type: essential
  points: 100
- name: volume_hostpath_not_found
  status: passed
  type: essential
  points: 100
- name: node_drain
  status: passed
  type: essential
  points: 100
- name: privileged_containers
  status: passed
  type: essential
  points: 100
- name: non_root_containers
  status: failed
  type: essential
  points: 0
- name: cpu_limits
  status: passed
  type: essential
  points: 100
- name: memory_limits
  status: passed
  type: essential
  points: 100
- name: hostpath_mounts
  status: passed
  type: essential
  points: 100
- name: container_sock_mounts
  status: passed
  type: essential
  points: 100
- name: selinux_options
  status: na
  type: essential
  points: 0
- name: hostport_not_used
  status: passed
  type: essential
  points: 100
- name: hardcoded_ip_addresses_in_k8s_runtime_configuration
  status: passed
  type: essential
  points: 100
- name: latest_tag
  status: passed
  type: essential
  points: 100
- name: log_output
  status: passed
  type: essential
  points: 100
- name: specialized_init_system
  status: failed
  type: essential
  points: 0
`

File: src/tasks/workload/compatibility.cr (hunk 5 above)
Comment (kosstennbl) [reply src/tasks/workload/compatibility.cr]: Done

File: src/tasks/platform/resilience.cr (hunk 4 above)
Comment (kosstennbl) [reply src/tasks/platform/resilience.cr]: Done

File: src/tasks/platform/resilience.cr (hunk 3 above)
Comment (kosstennbl) [reply src/tasks/platform/resilience.cr]: Done

File: src/tasks/platform/resilience.cr (hunk 2 above)
Comment (kosstennbl) [reply src/tasks/platform/resilience.cr]: Done

File: spec/platform/observability_spec.cr [hunk 6]
```diff
@@ -32,13 +32,10 @@ describe "Platform Observability" do
       Helm.helm_repo_add("prometheus-community","https://prometheus-community.github.io/helm-charts")
 		  result = ShellCmd.run("#{helm} install node-exporter prometheus-community/prometheus-node-exporter", force_output: true)
 
-      pod_ready = ""
-      pod_ready_timeout = 45
-      until (pod_ready == "true" || pod_ready_timeout == 0)
+      repeat_with_timeout(timeout: POD_READINESS_TIMEOUT, errormsg: "Pod readiness has timed-out") do
         pod_ready = KubectlClient::Get.pod_status("node-exporter-prometheus").split(",")[2]
```
[the remark this answers is not in the document]
Comment (kosstennbl) [reply spec/platform/observability_spec.cr]: Done, it wasn't that clear as, seemingly, values are not only "true" and "false". But as we're only interested of eqvivalence of the value to "true" or "True" (for node_ready) - changing this to boolean should be fine

File: src/tasks/utils/timeouts.cr [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    if result.nil?
+      if reset_on_nil
+        start_time = Time.utc
+      else
+        raise "Unexpected nil result of executed block, check the return value or parameter 'reset_on_nil'"
+      end
+    elsif result
+      return true
+    end
+    sleep 2
```
[the remark this answers is not in the document]
Comment (kosstennbl) [reply src/tasks/utils/timeouts.cr]: Done

**Comment by kosstennbl** [comment]
> Additionally, it would be good to document new environment vars to USAGE.md

Done

**Review by martin-mat — APPROVED** [review]
lgtm. Cool change, this helps to make the code more structured.

**Review by martin-mat — CHANGES_REQUESTED** [review]
It looks that spec tests are impacted by this change. Please analyze.
https://github.com/cnti-testcatalog/testsuite/actions/runs/9413816510

**Comment by kosstennbl** [comment]
Rebased, pushed fix for increase_decrease_capacity issue, shared_database2 seems to be an issue that is not caused by timeout redesign.

**Comment by martin-mat** [comment]
verified by
https://github.com/cnti-testcatalog/testsuite/actions/runs/9580002143

**Review by martin-mat — APPROVED** [review]
lgtm

**Pull request merged**